File: VB/DiagramDataControllerBehavior/Selectors.vb
Imports DevExpress.Diagram.Core
Imports System.Collections.Generic

Namespace DiagramDataControllerBehavior

    Public Class ItemsSelector
        Implements IChildrenSelector

        Public Function GetChildren(ByVal parent As Object) As IEnumerable(Of Object) Implements IChildrenSelector.GetChildren
            If TypeOf parent Is Department Then
                Return CType(parent, Department).Sections
            ElseIf TypeOf parent Is Section Then
                Return CType(parent, Section).Employees
            End If

            Return Nothing
        End Function
    End Class

    Public Class KeySelector
        Implements IKeySelector

        Private Function GetKey(ByVal obj As Object) As Object Implements IKeySelector.GetKey
            If TypeOf obj Is Department Then
                Return CType(obj, Department).DepartmentID
            ElseIf TypeOf obj Is Section Then
                Return CType(obj, Section).SectionID
            ElseIf TypeOf obj Is Employee Then
                Return CType(obj, Employee).EmployeeID
            End If

            Return obj
        End Function
    End Class
End Namespace

File: CS/DiagramDataControllerBehavior/Selectors.cs
using DevExpress.Diagram.Core;
using System.Collections.Generic;

namespace DiagramDataControllerBehavior {
    public class ItemsSelector : IChildrenSelector {
        public IEnumerable<object> GetChildren(object parent) {
            if (parent is Department)
                return ((Department)parent).Sections;
            else if (parent is Section)
                return ((Section)parent).Employees;
            return null;
        }
    }
    public class KeySelector : IKeySelector {
        object IKeySelector.GetKey(object obj) {
            if (obj is Department)
                return ((Department)obj).DepartmentID;
            else if (obj is Section)
                return ((Section)obj).SectionID;
            else if (obj is Employee)
                return ((Employee)obj).EmployeeID;
            return obj;
        }
    }
}
